File: challenges/realm/leetcode/linked_list_cycle.py
"""
Solution for LC#141: Linked List Cycle

https://leetcode.com/problems/linked-list-cycle/
"""
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def move(node, steps=1):
    if node is None:
        return None

    node = node.next
    if steps == 1:
        return node

    if node is not None:
        return node.next

    return None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False

        slow = move(head, steps=1)
        fast = move(head, steps=2)

        while fast is not None:  # if fast reaches the end (is None) there's no cycle!
            slow = move(slow, steps=1)
            fast = move(fast, steps=2)

            if slow == fast:  # slow has caught up with fast: cycle!
                return True

        return False

File: challenges/leetcode/linked_list_cycle.cpp
/*
 * Solution for LC#141: Linked List Cycle
 *
 * https://leetcode.com/problems/linked-list-cycle/
 */
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head ? head->next : nullptr;
        bool has_cycle = false;

        // breaks if the list reaches an end (fast == nullptr)
        // or if fast catches slow in a cycle (has_cycle == true)
        while (fast and !has_cycle) {
            slow = slow->next;
            fast = fast->next ? fast->next->next : nullptr;
            has_cycle = slow == fast;
        }

        return has_cycle;
    }
};
